File: appleApp/ui/portfolio/PortfolioScreen.swift
import Common
import SwiftUI

struct PortfolioScreen: View {
    private let coordinator: PortfolioCoordinator
    @State private var isSearchEnabled: Bool
    @State private var currentRoute: PortfolioRoute
    @State private var portfolioState: PortfolioState
    @State private var stopObserving: (() -> Void)?
    @State private var stopObservingPortfolio: (() -> Void)?

    init(coordinator: PortfolioCoordinator) {
        self.coordinator = coordinator
        _isSearchEnabled = State(initialValue:
            (coordinator.state.value as? PortfolioNavigationState)?.isSearchEnabled ?? false)
        _currentRoute = State(
            initialValue: (coordinator.state.value as? PortfolioNavigationState)?.currentRoute
                ?? .portfolio
        )
        _portfolioState = State(
            initialValue: coordinator.portfolioViewModel.state.value as! PortfolioState
        )
    }

    var body: some View {
        Group {
            if currentRoute == .portfolio {
                portfolioContent
            } else if currentRoute == .assetSearch {
                assetSearchContent
            } else if currentRoute == .edit {
                if let viewModel = coordinator.editViewModel {
                    EditScreen(viewModel: viewModel)
                } else {
                    portfolioContent
                }
            }
        }
        .onAppear {
            stopObserving?()
            stopObserving = coordinator.observeState {
                currentRoute = $0.currentRoute
                isSearchEnabled = $0.isSearchEnabled
            }
            stopObservingPortfolio?()
            stopObservingPortfolio = coordinator.portfolioViewModel.observeState {
                portfolioState = $0
            }
        }
        .onDisappear {
            stopObserving?()
            stopObserving = nil
            stopObservingPortfolio?()
            stopObservingPortfolio = nil
        }
    }

    private var portfolioContent: some View {
        ZStack(alignment: .bottomTrailing) {
            VStack(alignment: .leading, spacing: 0) {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Стоимость портфеля")
                        .font(.headline)
                        .foregroundStyle(.secondary)
                    Text("$\(portfolioState.totalValueUsdText)")
                        .font(.largeTitle.weight(.semibold))
                        .lineLimit(1)
                        .minimumScaleFactor(0.5)
                }
                .padding(.horizontal, 16)
                .padding(.vertical, 20)

                Divider()

                if portfolioState.isLoading && portfolioState.positions.isEmpty {
                    ProgressView()
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                } else if portfolioState.storageFailure != nil && portfolioState.positions.isEmpty {
                    VStack(spacing: 12) {
                        Text("Не удалось загрузить портфель")
                        Button("Повторить") { coordinator.portfolioViewModel.loadPositions() }
                    }
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                } else if portfolioState.rows.isEmpty {
                    Text("Добавьте первый актив")
                        .foregroundStyle(.secondary)
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                } else {
                    List(portfolioState.rows, id: \.position.asset.searchId) { row in
                        PortfolioRow(
                            row: row,
                            logoURL: coordinator.portfolioViewModel.logoUrl(asset: row.position.asset)
                        )
                    }
                    .listStyle(.plain)
                }
            }

            Button {
                coordinator.portfolioViewModel.onAssetSearch()
                syncRoute()
            } label: {
                Image(systemName: "plus")
                    .font(.title2)
                    .frame(width: 56, height: 56)
            }
            .disabled(!isSearchEnabled)
            .buttonStyle(.borderedProminent)
            .clipShape(Circle())
            .accessibilityLabel("Добавить актив")
            .padding(24)
        }
    }

    private var assetSearchContent: some View {
        VStack(alignment: .leading) {
            Button {
                coordinator.assetSearchViewModel.onBack()
                syncRoute()
            } label: {
                Label("Назад", systemImage: "chevron.left")
            }

            AssetSearchScreen(viewModel: coordinator.assetSearchViewModel)
        }
        .padding(16)
    }

    private func syncRoute() {
        currentRoute = (coordinator.state.value as? PortfolioNavigationState)?.currentRoute
            ?? .portfolio
    }
}

private struct PortfolioRow: View {
    let row: PortfolioRowState
    let logoURL: String?

    var body: some View {
        HStack(spacing: 12) {
            AssetLogo(url: logoURL)
            VStack(alignment: .leading, spacing: 4) {
                Text(row.position.asset.ticker).font(.headline)
                Text("\(row.position.quantity.value) \(row.position.asset.ticker)")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            }
            Spacer(minLength: 8)
            Text("$\(row.valueUsdText)")
                .font(.headline)
                .lineLimit(1)
                .minimumScaleFactor(0.7)
        }
        .padding(.vertical, 4)
    }
}

File: androidApp/src/main/kotlin/com/crossfolio/android/ui/portfolio/PortfolioScreen.kt
package com.crossfolio.android.ui.portfolio

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FilledIconButton
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.crossfolio.android.ui.portfolio.assetsearch.AssetSearchScreen
import com.crossfolio.android.ui.portfolio.edit.EditScreen
import com.crossfolio.common.portfolio.PortfolioCoordinator
import com.crossfolio.common.portfolio.PortfolioRoute
import com.crossfolio.common.portfolio.overview.PortfolioViewModel
import com.crossfolio.common.portfolio.overview.PortfolioRowState

@Composable
fun PortfolioScreen(coordinator: PortfolioCoordinator) {
    val state by coordinator.state.collectAsState()

    when (state.currentRoute) {
        PortfolioRoute.PORTFOLIO -> PortfolioContent(coordinator.portfolioViewModel, state.isSearchEnabled)
        PortfolioRoute.ASSET_SEARCH -> AssetSearchScreen(coordinator.assetSearchViewModel)
        PortfolioRoute.EDIT -> coordinator.editViewModel?.let { EditScreen(it) }
    }
}

@Composable
private fun PortfolioContent(viewModel: PortfolioViewModel, isSearchEnabled: Boolean) {
    val state by viewModel.state.collectAsState()
    LaunchedEffect(viewModel, isSearchEnabled) { viewModel.loadPositions() }

    Box(modifier = Modifier.fillMaxSize()) {
        Column(modifier = Modifier.fillMaxSize()) {
            Column(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 20.dp),
                verticalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                Text(
                    text = "Стоимость портфеля",
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Text(
                    text = "\$${state.totalValueUsdText}",
                    style = MaterialTheme.typography.headlineLarge,
                    fontWeight = FontWeight.SemiBold,
                )
            }
            HorizontalDivider()

            when {
                state.isLoading && state.positions.isEmpty() -> Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center,
                ) { CircularProgressIndicator() }
                state.storageFailure != null && state.positions.isEmpty() -> Column(
                    modifier = Modifier.fillMaxSize(),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.Center,
                ) {
                    Text("Не удалось загрузить портфель")
                    Button(onClick = viewModel::loadPositions) { Text("Повторить") }
                }
                state.rows.isEmpty() -> Box(
                    modifier = Modifier.fillMaxSize().padding(24.dp),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text = "Добавьте первый актив",
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
                else -> LazyColumn(
                    contentPadding = PaddingValues(bottom = 96.dp),
                ) {
                    items(
                        items = state.rows,
                        key = { "${it.position.asset.searchPlatform.name}:${it.position.asset.searchId}" },
                    ) { row ->
                        PortfolioRow(row, viewModel)
                        HorizontalDivider(modifier = Modifier.padding(start = 72.dp))
                    }
                }
            }
        }
        FilledIconButton(
            enabled = isSearchEnabled,
            onClick = viewModel::onAssetSearch,
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .padding(24.dp)
                .size(56.dp),
        ) {
            Text(text = "+", modifier = Modifier.semantics { contentDescription = "Добавить актив" })
        }
    }
}

@Composable
private fun PortfolioRow(row: PortfolioRowState, viewModel: PortfolioViewModel) {
    val position = row.position
    Row(
        modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        AssetLogo(viewModel.logoUrl(position.asset), viewModel::loadImage)
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(position.asset.ticker, style = MaterialTheme.typography.titleMedium)
            Text(
                text = "${position.quantity.value} ${position.asset.ticker}",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        Text(
            text = "\$${row.valueUsdText}",
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.Medium,
        )
    }
}
